### ISSUE TITLE
Cannot depend on internal library from tests

### ISSUE THREAD

**Issue by phadej** [haskell/cabal#6961]
```
Resolving dependencies...
cabal: Could not resolve dependencies:
[__0] trying: cabal-install-3.5.0.0 (user goal)
[__1] rejecting: cabal-install:!test (constraint from config file, command
line flag, or user target requires opposite flag selection)
[__1] rejecting: cabal-install:*test (requires library
'cabal-install-solver-dsl' from cabal-install, but the component is private)
[__1] fail (backjumping, conflict set: cabal-install, cabal-install:test)
After searching the rest of the dependency tree exhaustively, these were the
goals I've had most trouble fulfilling: cabal-install, cabal-install:test
```

with `cabal-install-3.4.0.0-rc1`

cc @grayjay @fgaz 

EDIT: for easy reproduction check https://github.com/phadej/cabal-fmt

```
% cabal build all --dry            
Resolving dependencies...
cabal: Could not resolve dependencies:
[__0] trying: cabal-fmt-0.1.3 (user goal)
[__1] rejecting: cabal-fmt:!test (constraint from config file, command line
flag, or user target requires opposite flag selection)
[__1] rejecting: cabal-fmt:*test (requires library 'cabal-fmt-internal' from
cabal-fmt, but the component is private)
[__1] fail (backjumping, conflict set: cabal-fmt, cabal-fmt:test)
After searching the rest of the dependency tree exhaustively, these were the
goals I've had most trouble fulfilling: cabal-fmt, cabal-fmt:test
```

If I comment a test-suite there, then the solver will find the solution, i.e. dependency from an executable in the same package is fine!

**Comment by phadej** [comment]
If this cannot be solved, I'm afraid we need to revert the solver patch for multilibs.

**Comment by grayjay** [comment]
I think the issue is that the solver merges all components within a package and doesn't try to handle intra-package dependencies, but the multilibs change checked the visibility for all dependencies.  The solver filters out intra-package dependencies in the build phase (https://github.com/haskell/cabal/blob/6f22f2a789fa23edb210a2591d74ea6a5f767872/cabal-install/Distribution/Solver/Modular/Builder.hs#L76-L85), but that doesn't affect the component check in Validate.hs.

@phadej Thanks for making a fix and adding more tests!  I think  #6964 will fix this issue, though a more general fix would be to filter out intra-package dependencies earlier in dependency solving.

**Comment by phadej** [comment]
@grayjay Do you mind opening an issue for 

>  a more general fix would be to filter out intra-package dependencies earlier in dependency solving.

I don't know how to do that, so going with #6964 for now. Thanks for prompt review!
